verification/src/solidity/compiler_fetcher.rs

```diff
@@ -0,0 +1,295 @@
+use crate::compiler::{
+    fetcher::Fetcher,
+    version::{CompilerVersion, ParseError},
+};
+use async_trait::async_trait;
+use serde::Deserialize;
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io::ErrorKind,
+    os::unix::prelude::OpenOptionsExt,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Error, Debug)]
+pub enum ListError {
+    #[error("fetching list json returned error: {0}")]
+    ListJsonFetch(reqwest::Error),
+    #[error("cannot parse list json file: {0}")]
+    ParseListJson(reqwest::Error),
+    #[error("error parsing tag name into version: {0}")]
+    TagName(ParseError),
+    #[error("error parsing 'path' field: {0}")]
+    Path(url::ParseError),
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct ListJson {
+    pub builds: Vec<CompilerInfo>,
+}
+
+impl ListJson {
+    fn into_releases(
+        self,
+        download_prefix: &Url,
+    ) -> Result<HashMap<CompilerVersion, Url>, ListError> {
+        let mut releases = HashMap::default();
+        for build in self.builds {
+            let version =
+                CompilerVersion::from_str(&build.long_version).map_err(ListError::TagName)?;
+            let download_url = build
+                .path
+                .try_into_url(download_prefix)
+                .map_err(ListError::Path)?;
+            releases.insert(version, download_url);
+        }
+        Ok(releases)
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct CompilerInfo {
+    pub path: DownloadPath,
```

Comment:
What if we would keep the final URL here (prepending paths without schema with url prefix on initialization)?